model/common/src/icon4py/model/common/states/factory.py

```diff
@@ -0,0 +1,476 @@
+        offset_providers = self._get_offset_providers(grid_provider.grid)
+        deps.update(dims)
+        self._func.with_backend(backend)(**deps, offset_provider=offset_providers)
+
+    @property
+    def fields(self) -> Mapping[str, state_utils.FieldType]:
+        return self._fields
+
+    @property
+    def func(self) -> Callable:
+        return self._func
+
+    @property
+    def dependencies(self) -> Sequence[str]:
+        return list(self._dependencies.values())
+
+
+class NumpyFieldsProvider(FieldProvider):
+    """
+    Computes a field defined by a numpy function.
+
+    TODO (halungge): need to specify a parameter source to be able to postpone evaluation:  paramters are mostly
+                    configuration values
+
+
+    Args:
+        func: numpy function that computes the fields
+        domain: the compute domain used for the stencil computation
+        fields: Seq[str] names under which the results fo the function will be registered
+        deps: dict[str, str] input fields used for computing this stencil: the key is the variable name
+            used in the program and the value the name of the field it depends on.
+        params: scalar arguments for the function
+    """
+
+    def __init__(
+        self,
+        func: Callable,
+        domain: dict[gtx.Dimension : tuple[DomainType, DomainType]],
+        fields: Sequence[str],
+        deps: dict[str, str],
+        offsets: Optional[dict[str, gtx.Dimension]] = None,
+        params: Optional[dict[str, state_utils.ScalarType]] = None,
+    ):
+        self._func = func
+        self._compute_domain = domain
+        self._offsets = offsets
+        self._dims = domain.keys()
+        self._fields: dict[str, Optional[state_utils.FieldType]] = {name: None for name in fields}
+        self._dependencies = deps
+        self._offsets = offsets if offsets is not None else {}
+        self._params = params if params is not None else {}
+
+    def __call__(
+        self,
+        field_name: str,
+        factory: FieldSource,
+        backend: gtx_backend.Backend,
+        grid: GridProvider,
+    ) -> state_utils.FieldType:
+        if any([f is None for f in self.fields.values()]):
+            self._compute(factory, backend, grid)
+        return self.fields[field_name]
+
+    def _compute(
+        self,
+        factory: FieldSource,
+        backend: gtx_backend.Backend,
+        grid_provider: GridProvider,
+    ) -> None:
+        self._validate_dependencies()
+        args = {k: factory.get(v).ndarray for k, v in self._dependencies.items()}
+        offsets = {k: grid_provider.grid.connectivities[v] for k, v in self._offsets.items()}
+        args.update(offsets)
+        args.update(self._params)
+        results = self._func(**args)
+        ## TODO: can the order of return values be checked?
+        results = (results,) if isinstance(results, xp.ndarray) else results
```

Comment:
```suggestion
        ## TODO: can the order of return values be checked?
        results = (results,) if isinstance(self._func(**args), xp.ndarray) else results
```
Just as a personal preference